src/panels/ConfigPanel/ConfigPanel.tsx

```diff
@@ -48,21 +50,25 @@ export type ConfigProps = {
 
 type SetNewValue = (name: string, value: any) => void;
 
+type ConfirmTypes = "apply" | "cancel";
+
 export default function ConfigPanel({
   appName,
   charm,
   modelUUID,
-  closePanel,
+  onClose,
 }: Props): ReactElement {
   const reduxStore = useStore();
   const [config, setConfig] = useState<Config>({});
   const [selectedConfig, setSelectedConfig] = useState<ConfigData | undefined>(
     undefined
   );
-  const [enableSave, setEnableSave] = useState<Boolean>(false);
-  const [showResetAll, setShowResetAll] = useState<Boolean>(false);
-  const [isLoading, setIsLoading] = useState<Boolean>(true);
-  const [savingConfig, setSavingConfig] = useState<Boolean>(false);
+  const [enableSave, setEnableSave] = useState<boolean>(false);
+  const [showResetAll, setShowResetAll] = useState<boolean>(false);
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+  const [savingConfig, setSavingConfig] = useState<boolean>(false);
+  const [confirmOpen, setConfirmOpen] = useState<boolean>(false);
+  const [confirmType, setConfirmType] = useState<ConfirmTypes>("apply");
```

Comment:
Merging these values into one 'confirm' object and updating it once would reduce rerendering and eliminate possible race conditons.

Reply:
Good call.